sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala

```diff
@@ -1657,6 +1656,28 @@ class InsertSuite extends DataSourceTest with SharedSparkSession {
     }
   }
 
+  test("SPARK-40001 JSON DEFAULT columns require JSON_GENERATOR_IGNORE_NULL_FIELDS off") {
+    val error = "DEFAULT values are not supported for JSON tables"
+    withTable("t") {
+      assert(intercept[AnalysisException] {
+        sql("create table t (a int default 42) using json")
```

Comment:
Hm, should we take `ignoreNullFields` option into account too? e.g.( create table t ... using json options (ignoreNullFields ... )

Reply:
Thanks for the useful comment! This made me realize that other writers may create JSON with missing NULL values in the storage as well. I fixed this by:

1) Reverting the analyzer changes in this PR
2) Changed the new config to `DEFAULT_COLUMN_JSON_GENERATOR_FORCE_NULL_FIELDS` which overrides any other settings for target columns with DEFAULT values to always write explicit NULLs to storage.
3) Updated the `DEFAULT_COLUMN_ALLOWED_PROVIDERS` config to ban `ALTER TABLE ADD COLUMN` commands with `DEFAULT` values for JSON tables, with a descriptive error message.

This ensures correctness with JSON `DEFAULT` columns by always ensuring that new rows with NULL values get explicit NULLs written to the JSON storage, so that subsequent scans can tell the difference between those NULLs and any written `DEFAULT` values.